scipy/stats/_morestats.py

```diff
@@ -283,23 +284,25 @@ def kstat(data, n=2):
     0.00166 0.00166 -4.99e-09
     -2.88e-06 -2.88e-06 8.63e-13
     """
+    xp = array_namespace(data)
+    data = xp.asarray(data)
     if n > 4 or n < 1:
         raise ValueError("k-statistics only supported for 1<=n<=4")
     n = int(n)
-    S = np.zeros(n + 1, np.float64)
-    data = ravel(data)
-    N = data.size
+    S = xp.zeros(n + 1, dtype=xp.float64)
+    data = xp.reshape(data, (-1,))
+    N = data.shape[0]
 
     # raise ValueError on empty input
     if N == 0:
         raise ValueError("Data input must not be empty")
 
     # on nan input, return nan without warning
-    if np.isnan(np.sum(data)):
-        return np.nan
+    if xp.isnan(xp.sum(data)):
+        return xp.nan
```

Comment:
Still need to figure out this failure
```
____________________________________________________________________________________ TestKstat.test_nan_input[array_api_strict] _____________________________________________________________________________________
scipy/stats/tests/test_morestats.py:1689: in test_nan_input
    xp_assert_equal(stats.kstat(data), xp.asarray(xp.nan))
E   AssertionError: Namespaces do not match.
E   Actual: scipy._lib.array_api_compat.numpy
E   Desired: array_api_strict
        data       = Array([ 0.,  1.,  2.,  3.,  4.,  5., nan,
        7.,  8.,  9.], dtype=array_api_strict.float64)
        self       = <scipy.stats.tests.test_morestats.TestKstat object at 0x7f5b96eb19d0>
        xp         = <module 'array_api_strict' from '/home/jakeb/miniconda3/envs/scipy-dev-pytorch/lib/python3.11/site-packages/array_api_strict/__init__.py'>
__________________________________________________________________________________________ TestKstat.test_nan_input[torch] __________________________________________________________________________________________
scipy/stats/tests/test_morestats.py:1689: in test_nan_input
    xp_assert_equal(stats.kstat(data), xp.asarray(xp.nan))
E   AssertionError: Namespaces do not match.
E   Actual: scipy._lib.array_api_compat.numpy
E   Desired: scipy._lib.array_api_compat.torch
        data       = tensor([0., 1., 2., 3., 4., 5., nan, 7., 8., 9.])
        self       = <scipy.stats.tests.test_morestats.TestKstat object at 0x7f5b96eb1c50>
        xp         = <module 'torch' from '/home/jakeb/miniconda3/envs/scipy-dev-pytorch/lib/python3.11/site-packages/torch/__init__.py'>
============================================================================================== short test summary info ==============================================================================================
FAILED scipy/stats/tests/test_morestats.py::TestKstat::test_nan_input[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/stats/tests/test_morestats.py::TestKstat::test_nan_input[torch] - AssertionError: Namespaces do not match.
```

Reply:
Use `_get_nan` to get the right NaN type and dtype.

Reply:
Thanks thought that might be the case